In [ ]:

import pandas as pd
import numpy as np
import re
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as stats


In [ ]:
df_final_demo = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_demo.txt'
df_final_web_data_pt1 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_1.txt'
df_final_web_data_pt2 = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt'
df_final_experiment_client = r'https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_experiment_clients.txt'



In [ ]:
# Intenta leer el archivo
try:
    df_final_demo = pd.read_csv(df_final_demo, sep=',')  
    df_final_web_data_pt1= pd.read_csv(df_final_web_data_pt1, sep=',')
    df_final_web_data_pt2= pd.read_csv(df_final_web_data_pt2, sep=',')
    df_final_experiment_client= pd.read_csv(df_final_experiment_client, sep=',')
except Exception as e:
    print(f"Error al cargar el archivo: {e}")

In [ ]:
df_final_demo.columns = df_final_demo.columns.str.lower()
df_final_web_data_pt1.columns = df_final_web_data_pt1.columns.str.lower()
df_final_web_data_pt2.columns = df_final_web_data_pt2.columns.str.lower()
df_final_experiment_client.columns = df_final_experiment_client.columns.str.lower()


In [ ]:
final_demo = df_final_demo
final_demo

In [ ]:
final_demo = pd.merge(final_demo, df_final_experiment_client, on='client_id')
final_demo

In [ ]:
final_demo = final_demo.dropna()
final_demo

In [ ]:
df_final_web_data_pt1

In [ ]:
df_final_web_data_pt2

In [ ]:
final_web = pd.concat([df_final_web_data_pt1, df_final_web_data_pt2], ignore_index=True)
final_web

In [ ]:
final_web.isna()

In [ ]:
print(final_web.isnull().sum())


In [ ]:
tabla_analisis = pd.merge(final_demo, final_web, on='client_id')

In [ ]:
tabla_analisis

In [ ]:
tabla_analisis.head()

In [ ]:
tabla_analisis["process_step"].unique()

In [ ]:
tabla_test = tabla_analisis.set_index(['client_id', 'process_step'])
tabla_test.tail(10)

In [ ]:
tabla_test

In [ ]:
tabla_analisis_reducida = tabla_analisis[['client_id', 'process_step']]
tabla_analisis_reducida

In [ ]:
pivot_df = tabla_analisis.pivot_table(index='client_id', columns='process_step', aggfunc='size', fill_value=0)
pivot_df

In [ ]:
clients_at_start = pivot_df[pivot_df['start'] > 0]
clients_at_start

In [ ]:
clients_at_step_1 = pivot_df[pivot_df['step_1'] > 0]
clients_at_step_1

In [ ]:
clients_at_step_2 = pivot_df[pivot_df['step_2'] > 0]
clients_at_step_2

In [ ]:
clients_at_step_3 = pivot_df[pivot_df['step_3'] > 0]
clients_at_step_3

In [ ]:
# Filtrar las filas donde 'process_step' es igual a 'confirm'
clients_at_confirm = pivot_df[pivot_df['confirm'] > 0]
clients_at_confirm


In [ ]:
confirm = len(clients_at_confirm)
start = len(clients_at_start)
step1 = len(clients_at_step_1)
step2 = len(clients_at_step_2)
step3 = len(clients_at_step_3)
total_usuarios = len(df_final_demo)
total_test = len(pivot_df)

In [ ]:
total_usuarios, total_test, start ,step1, step2 ,step3 , confirm

In [ ]:
d_index = {
    "name": ['total_usuarios', 'total_test', 'start' ,'step1', 'step2' ,'step3' , 'confirm'],
    "score": [total_usuarios, total_test, start ,step1, step2 ,step3 , confirm]
}

df_index = pd.DataFrame(d_index)
df_index

TOTAL CLIENTES= 70,609
START = 50,063
STEP 1 = 44,406
STEP 2 = 40,896
STEP 3 = 38,292
CONFIRM = 34,110
% de caída de START a STEP 1: 11.37%
% de caída de STEP 1 a STEP 2: 7.97%
% de caída de STEP 2 a STEP 3: 6.37%
% de caída de STEP 3 a CONFIRM: 10.92%
% de clientes que han confirmado: 48.3%

In [ ]:
pivot_df.sum()

In [ ]:
tabla_analisis.describe()

Answer the following questions about demographics:

Who are the primary clients using this online process?
Are the primary clients younger or older, new or long-standing?

In [ ]:
#Are the primary clients younger or older

# Calculate the mean age of the clients
mean_age = tabla_analisis['clnt_age'].mean()

# Print the mean age with 2 decimal places
print(f"La media de edad es de {mean_age:.2f}")

# Calculate the median tenure of the clients
median_age = tabla_analisis['clnt_age'].median()

# Print the median tenure with 2 decimal places
print(f"La mediana de edad es de {median_age:.2f}")

In [ ]:
#New or long-standing?

# Calculate the mean age of the clients
mean_standing_time = tabla_analisis['clnt_tenure_mnth'].mean()

# Print the mean age with 2 decimal places
print(f"La media de meses de estancia en la plataforma es de {mean_standing_time:.2f}")

# Calculate the median tenure of the clients
median_standing_time = tabla_analisis['clnt_tenure_mnth'].median()

# Print the median tenure with 2 decimal places
print(f"La mediana de meses de estancia en la plataforma es de {median_standing_time:.2f}")


In [ ]:
tabla_analisis.head()

In [ ]:
# Filter the DataFrame to include only rows where 'client_id' is 836976
client_rows = tabla_analisis[tabla_analisis['client_id'] == 1516]

# Display the filtered rows
client_rows

In [ ]:
# Count the number of occurrences of each unique value in the 'gendr' column
gender_counts = tabla_analisis['gendr'].value_counts()

# Print the counts of each gender
print(gender_counts)

In [ ]:
#Completion Rate: The proportion of users who reach the final ‘confirm’ step.

completion_counts = tabla_analisis['process_step'].value_counts()
print(completion_counts)

In [ ]:
tabla_analisis['variation'].value_counts()

In [ ]:
# Filter the DataFrame to include only rows where the step is 'confirm'
confirm_clients = tabla_analisis[tabla_analisis['process_step'] == 'confirm']

# Get the unique client IDs who reached the 'confirm' step
unique_confirm_clients = confirm_clients['client_id'].unique()

In [ ]:
len(unique_confirm_clients)

In [ ]:
unique_clients_at_confirm=(len(unique_confirm_clients)/len(tabla_analisis['client_id'].unique()))*100
print(f"La tasa de finalización es de {unique_clients_at_confirm:.2f}")

In [ ]:
confirm_clients_test = confirm_clients[confirm_clients['variation'] == 'Test']
confirm_clients_control = confirm_clients[confirm_clients['variation'] == 'Control']
len(confirm_clients_test), len(confirm_clients_control)

In [ ]:
# Get the unique client IDs who reached the 'confirm' step from each group
unique_confirm_clients_test = confirm_clients_test['client_id'].unique()
unique_confirm_clients_control = confirm_clients_control['client_id'].unique()
len(unique_confirm_clients_test), len(unique_confirm_clients_control)

Based on the chosen KPIs, how does the new design’s performance compare to the old one?

The new design (test) has more number confirms than the older (control)

(25716 VS 17498)

And the % is higher

variation
Test       177787
Control    143408
Name: count, dtype: int64

14.46% Test
12.20% Control

In [ ]:
tabla_analisis.dtypes

In [ ]:
# Convert the timestamp column to datetime
tabla_analisis['date_time'] = pd.to_datetime(tabla_analisis['date_time'])

# Sort the DataFrame by user_id and timestamp
user_steps = tabla_analisis.sort_values(by=['client_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps['date_time'] = user_steps.groupby('client_id')['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step:")
print(average_time_spent)

In [ ]:
# Calculate the average time spent on each step
median_time_spent = user_steps.groupby('process_step')['date_time'].median()

# Print the average time spent on each step
print("Median Value of Time Spent on Each Step:")
print(median_time_spent)

In [ ]:
# Sort the DataFrame by user_id and timestamp
user_steps_visit_id = tabla_analisis.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps_visit_id['date_time'] = user_steps_visit_id.groupby('visit_id')['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_visit_id.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step:")
print(average_time_spent)

In [ ]:
print(user_steps_visit_id.head(100))

In [ ]:
# Display the first 100 rows of the user_steps DataFrame
user_steps.head(10)

In [ ]:
tabla_analisis.head()

In [ ]:
# Filter the DataFrame to include only rows where 'client_id' is 836976
client_rows = user_steps[user_steps['client_id'] == 1516]

# Display the filtered rows
client_rows

In [191]:
test_group = tabla_analisis[tabla_analisis['variation'] == 'Test']
control_group = tabla_analisis[tabla_analisis['variation'] == 'Control']

In [185]:
test_group, control_group

(        client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
 0          836976             6.0              73.0      60.5     U   
 1          836976             6.0              73.0      60.5     U   
 2          836976             6.0              73.0      60.5     U   
 3          836976             6.0              73.0      60.5     U   
 4          836976             6.0              73.0      60.5     U   
 ...           ...             ...               ...       ...   ...   
 321190    7468138            18.0             222.0      61.0     F   
 321191    7468138            18.0             222.0      61.0     F   
 321192    7468138            18.0             222.0      61.0     F   
 321193    7468138            18.0             222.0      61.0     F   
 321194    7468138            18.0             222.0      61.0     F   
 
         num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
 0             2.0   45105.30           6.0            9.0    

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
user_steps_test = test_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Calculate the time spent on each step by each user within each visit
user_steps_test['time_spent'] = user_steps_test.groupby(['client_id', 'visit_id'])['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_test.groupby('process_step')['time_spent'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step for Test group:")
print(average_time_spent)

In [ ]:
# Sort the DataFrame by user_id and timestamp
user_steps_control = control_group.sort_values(by=['client_id', 'date_time'])

# Calculate the time spent on each step by each user
user_steps_control['date_time'] = user_steps_control.groupby(['client_id', 'visit_id'])['date_time'].diff().shift(-1)

# Calculate the average time spent on each step
average_time_spent = user_steps_control.groupby('process_step')['date_time'].mean()

# Print the average time spent on each step
print("Average Time Spent on Each Step for Control group:")
print(average_time_spent)

In [ ]:
# Filter the DataFrame to include only rows where time spent is greater than 10 hours
ten_hours = pd.Timedelta(hours=10)
filtered_users = user_steps[user_steps['date_time'] > ten_hours]

# Display the filtered rows
len(filtered_users['client_id'].unique())

In [ ]:
# Filter the DataFrame to include only rows where time spent is greater than 10 hours
one_hour = pd.Timedelta(hours=1)
filtered_users = user_steps[user_steps['date_time'] < one_hour]

# Display the filtered rows
len(filtered_users['client_id'].unique())

In [ ]:
tabla_analisis.head()

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted = tabla_analisis.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted['step_order'] = df_sorted['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted['previous_step_order'] = df_sorted.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted['error'] = df_sorted['step_order'] < df_sorted['previous_step_order']

# Count the number of errors
error_count = df_sorted['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted['error'].count() - df_sorted.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:
# Filter the DataFrame to show only rows where 'error' is True
errors_df = df_sorted[df_sorted['error'] == True]

# Display the filtered DataFrame
errors_df[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:
# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted_test = test_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted_test['step_order'] = df_sorted_test['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted_test['previous_step_order'] = df_sorted_test.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted_test['error'] = df_sorted_test['step_order'] < df_sorted_test['previous_step_order']

# Count the number of errors
error_count = df_sorted_test['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted_test['error'].count() - df_sorted_test.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted_test[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

In [ ]:

# Sort the DataFrame by 'client_id', 'visit_id', and 'date_time'
df_sorted_control = control_group.sort_values(by=['client_id', 'visit_id', 'date_time'])

# Define the order of steps
step_order = {'start': 1, 'step_1': 2, 'step_2': 3, 'step_3': 4, 'confirm': 5}

# Map the process steps to their order
df_sorted_control['step_order'] = df_sorted_control['process_step'].map(step_order)

# Identify instances where users move from a later step to an earlier step
df_sorted_control['previous_step_order'] = df_sorted_control.groupby(['client_id', 'visit_id'])['step_order'].shift(1)
df_sorted_control['error'] = df_sorted_control['step_order'] < df_sorted_control['previous_step_order']

# Count the number of errors
error_count = df_sorted_control['error'].sum()

# Calculate the total number of transitions
total_transitions = df_sorted_control['error'].count() - df_sorted_control.groupby(['client_id', 'visit_id']).ngroups

# Calculate the error rate
error_rate = error_count / total_transitions

# Display the error rate
print(f"Error Rate: {error_rate:.2%}")

# Display the DataFrame with error identification
df_sorted_control[['client_id', 'visit_id', 'process_step', 'date_time', 'step_order', 'previous_step_order', 'error']]

H4.

The new design (test) has more number confirms than the older (control)

(Test 25716 VS Control 17498)

And the % is higher

variation
Test       177787
Control    143408
Name: count, dtype: int64

14.46% Test
12.20% Control

To conduct hypothesis testing on the completion rate between the Test and Control groups, follow these steps:

Define the Hypotheses:

Null Hypothesis (H0): There is no difference in the completion rates between the Test and Control groups.
Alternative Hypothesis (H1): There is a difference in the completion rates between the Test and Control groups.
Choose the Significance Level:

Commonly used significance level (α) is 0.05.
Calculate the Completion Rates:

Calculate the completion rates for both the Test and Control groups.
Perform the Statistical Test:

Use a two-proportion z-test to compare the completion rates between the two groups.
Interpret the Results:

Determine if you can reject the null hypothesis based on the p-value and the significance level.

In [175]:
import scipy.stats as st
from scipy.stats import ttest_ind

In [182]:
tabla_analisis

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0          836976             6.0              73.0      60.5     U   
1          836976             6.0              73.0      60.5     U   
2          836976             6.0              73.0      60.5     U   
3          836976             6.0              73.0      60.5     U   
4          836976             6.0              73.0      60.5     U   
...           ...             ...               ...       ...   ...   
321190    7468138            18.0             222.0      61.0     F   
321191    7468138            18.0             222.0      61.0     F   
321192    7468138            18.0             222.0      61.0     F   
321193    7468138            18.0             222.0      61.0     F   
321194    7468138            18.0             222.0      61.0     F   

        num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0   45105.30           6.0            9.0      Test   
1             2.0   45105.30           6.0            9.0      Test   
2             2.0   45105.30           6.0            9.0      Test   
3             2.0   45105.30           6.0            9.0      Test   
4             2.0   45105.30           6.0            9.0      Test   
...           ...        ...           ...            ...       ...   
321190        3.0  209278.15           0.0            3.0      Test   
321191        3.0  209278.15           0.0            3.0      Test   
321192        3.0  209278.15           0.0            3.0      Test   
321193        3.0  209278.15           0.0            3.0      Test   
321194        3.0  209278.15           0.0            3.0      Test   

                   visitor_id                      visit_id process_step  \
0        427070339_1413275162   228976764_46825473280_96584      confirm   
1        427070339_1413275162   228976764_46825473280_96584      confirm   
2        427070339_1413275162   228976764_46825473280_96584      confirm   
3        427070339_1413275162   228976764_46825473280_96584       step_3   
4        427070339_1413275162   228976764_46825473280_96584       step_2   
...                       ...                           ...          ...   
321190  317123960_46820919455  769876461_30381166055_830233       step_2   
321191  317123960_46820919455  769876461_30381166055_830233       step_1   
321192  317123960_46820919455  769876461_30381166055_830233        start   
321193  317123960_46820919455  769876461_30381166055_830233        start   
321194  317123960_46820919455  769876461_30381166055_830233        start   

                 date_time  
0      2017-04-02 11:51:13  
1      2017-04-02 11:47:50  
2      2017-04-02 11:46:45  
3      2017-04-02 11:23:08  
4      2017-04-02 11:22:24  
...                    ...  
321190 2017-03-30 23:59:15  
321191 2017-03-30 23:58:51  
321192 2017-03-30 23:58:40  
321193 2017-03-30 23:55:11  
321194 2017-03-31 00:16:12  

[321195 rows x 14 columns]

In [186]:
test_group

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0          836976             6.0              73.0      60.5     U   
1          836976             6.0              73.0      60.5     U   
2          836976             6.0              73.0      60.5     U   
3          836976             6.0              73.0      60.5     U   
4          836976             6.0              73.0      60.5     U   
...           ...             ...               ...       ...   ...   
321190    7468138            18.0             222.0      61.0     F   
321191    7468138            18.0             222.0      61.0     F   
321192    7468138            18.0             222.0      61.0     F   
321193    7468138            18.0             222.0      61.0     F   
321194    7468138            18.0             222.0      61.0     F   

        num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0   45105.30           6.0            9.0      Test   
1             2.0   45105.30           6.0            9.0      Test   
2             2.0   45105.30           6.0            9.0      Test   
3             2.0   45105.30           6.0            9.0      Test   
4             2.0   45105.30           6.0            9.0      Test   
...           ...        ...           ...            ...       ...   
321190        3.0  209278.15           0.0            3.0      Test   
321191        3.0  209278.15           0.0            3.0      Test   
321192        3.0  209278.15           0.0            3.0      Test   
321193        3.0  209278.15           0.0            3.0      Test   
321194        3.0  209278.15           0.0            3.0      Test   

                   visitor_id                      visit_id process_step  \
0        427070339_1413275162   228976764_46825473280_96584      confirm   
1        427070339_1413275162   228976764_46825473280_96584      confirm   
2        427070339_1413275162   228976764_46825473280_96584      confirm   
3        427070339_1413275162   228976764_46825473280_96584       step_3   
4        427070339_1413275162   228976764_46825473280_96584       step_2   
...                       ...                           ...          ...   
321190  317123960_46820919455  769876461_30381166055_830233       step_2   
321191  317123960_46820919455  769876461_30381166055_830233       step_1   
321192  317123960_46820919455  769876461_30381166055_830233        start   
321193  317123960_46820919455  769876461_30381166055_830233        start   
321194  317123960_46820919455  769876461_30381166055_830233        start   

                 date_time  
0      2017-04-02 11:51:13  
1      2017-04-02 11:47:50  
2      2017-04-02 11:46:45  
3      2017-04-02 11:23:08  
4      2017-04-02 11:22:24  
...                    ...  
321190 2017-03-30 23:59:15  
321191 2017-03-30 23:58:51  
321192 2017-03-30 23:58:40  
321193 2017-03-30 23:55:11  
321194 2017-03-31 00:16:12  

[177787 rows x 14 columns]

In [192]:
test_confirmed = test_group['process_step'].apply(lambda x: 1 if x == 'confirm' else 0)
control_confirmed = control_group['process_step'].apply(lambda x: 1 if x == 'confirm' else 0)

In [195]:
test_confirmed, control_confirmed

(0         1
 1         1
 2         1
 3         0
 4         0
          ..
 321190    0
 321191    0
 321192    0
 321193    0
 321194    0
 Name: process_step, Length: 177787, dtype: int64,
 11        1
 12        0
 13        0
 14        0
 15        0
          ..
 321169    1
 321170    0
 321171    0
 321172    0
 321173    0
 Name: process_step, Length: 143408, dtype: int64)

In [194]:
# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_confirmed, control_confirmed, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in completion rates between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in completion rates between the Test and Control groups.")

Test Statistic: 18.83876734309717
P-value: 4.0175581234786464e-79
Reject the null hypothesis: There is a significant difference in completion rates between the Test and Control groups.


### Conclusion:

Based on the results of the Two-sample t-test:

- **Test Statistic**: 18.83876734309717
- **P-value**: 4.0175581234786464e-79

Since the p-value is significantly less than the commonly used significance level of 0.05, we reject the null hypothesis.

### Interpretation:
- **Reject the Null Hypothesis**: There is a statistically significant difference in completion rates between the Test and Control groups.
- **Significance**: The extremely low p-value indicates that the observed difference in completion rates is highly unlikely to have occurred by chance.

### Practical Implications:
- **Effectiveness of the Redesign**: The new design (Test group) has a significantly higher completion rate compared to the old design (Control group).
- **Decision Making**: Based on this analysis, it is recommended to adopt the new design, as it leads to a higher completion rate, which can positively impact user engagement and overall performance metrics.

### Summary:
The hypothesis testing confirms that the new design is more effective in terms of completion rates. The significant difference observed suggests that the redesign has successfully improved user experience and completion rates. Therefore, the new design should be implemented to achieve better outcomes.

H5.

Completion Rate with a Cost-Effectiveness Threshold
The introduction of a new UI design comes with its associated costs: design, development, testing, potential training for staff, and possible short-term disruptions or adjustments for users. To justify these costs, Vanguard has determined that any new design should lead to a minimum increase in the completion rate to be deemed cost-effective.

Threshold: Vanguard has set this minimum increase in completion rate at 5%. This is the rate at which the projected benefits, in terms of increased user engagement and potential revenue, are estimated to outweigh the costs of the new design.

You are required to carry out another analysis, ensuring that the observed increase in completion rate from the A/B test meets or exceeds this 5% threshold. If the new design doesn’t lead to at least this level of improvement, it may not be justifiable from a cost perspective, regardless of its statistical significance.

In [ ]:
from scipy.stats import norm
from statsmodels.stats.proportion import proportions_ztest

# Calculate the number of completions and total observations for each group
completion_counts = df.groupby('group')['completed'].sum()
total_counts = df.groupby('group')['completed'].count()

# Calculate completion rates
completion_rate_test = completion_counts['Test'] / total_counts['Test']
completion_rate_control = completion_counts['Control'] / total_counts['Control']

# Calculate the observed increase in completion rate
observed_increase = completion_rate_test - completion_rate_control

# Define the threshold
threshold = 0.05

# Perform the one-sided two-proportion z-test
count = completion_counts.values
nobs = total_counts.values
stat, p_value = proportions_ztest(count, nobs, alternative='larger')

# Define the significance level
alpha = 0.05

# Print the results
print(f"Completion Rate (Test): {completion_rate_test:.4f}")
print(f"Completion Rate (Control): {completion_rate_control:.4f}")
print(f"Observed Increase: {observed_increase:.4f}")
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if observed_increase >= threshold and p_value < alpha:
    print("Reject the null hypothesis: The observed increase in completion rate meets or exceeds the 5% threshold and is statistically significant.")
else:
    print("Fail to reject the null hypothesis: The observed increase in completion rate does not meet the 5% threshold or is not statistically significant.")

Completion Rate (Test): 0.1446
Completion Rate (Control): 0.1220
Observed Increase: 0.0226
Test Statistic: -18.684265189201895
P-value: 1.0
Fail to reject the null hypothesis: The observed increase in completion rate does not meet the 5% threshold or is not statistically significant.


In [197]:
# Given data
test_completions = 25716
control_completions = 17498
test_total = 177787
control_total = 143408

# Calculate completion rates
completion_rate_test = test_completions / test_total
completion_rate_control = control_completions / control_total

# Calculate the observed increase in completion rate
observed_increase = completion_rate_test - completion_rate_control

# Define the threshold
threshold = 0.05

# Perform the one-sided two-proportion z-test
count = [test_completions, control_completions]
nobs = [test_total, control_total]
stat, p_value = proportions_ztest(count, nobs, alternative='larger')

# Define the significance level
alpha = 0.05

# Print the results
print(f"Completion Rate (Test): {completion_rate_test:.4f}")
print(f"Completion Rate (Control): {completion_rate_control:.4f}")
print(f"Observed Increase: {observed_increase:.4f}")
print(f"Test Statistic: {stat}")

Completion Rate (Test): 0.1446
Completion Rate (Control): 0.1220
Observed Increase: 0.0226
Test Statistic: 18.684265189201895


### Conclusions:

Based on the results of the one-sided two-proportion z-test:

- **Completion Rate (Test)**: 14.46%
- **Completion Rate (Control)**: 12.20%
- **Observed Increase**: 2.26%
- **Test Statistic**: -18.684265189201895
- **P-value**: 1.0

### Interpretation:
- **Fail to Reject the Null Hypothesis**: The observed increase in completion rate does not meet the 5% threshold and is not statistically significant.
- **Threshold Not Met**: The observed increase of 2.26% is below the required 5% threshold set by Vanguard to justify the costs associated with the new design.

### Practical Implications:
- **Cost-Effectiveness**: The new design does not lead to a sufficient increase in completion rate to be deemed cost-effective. The increase in completion rate is not large enough to outweigh the costs of the new design.
- **Decision Making**: Based on this analysis, it is not recommended to adopt the new design solely based on the completion rate improvement, as it does not meet the cost-effectiveness threshold.

### Summary:
The hypothesis testing confirms that the new design does not achieve the required 5% increase in completion rate to justify its costs. Although the new design has a higher completion rate compared to the old design, the increase is not statistically significant and does not meet the threshold set by Vanguard. Therefore, the new design may not be justifiable from a cost perspective.

H6.

You have been given the freedom to choose another hypothesis to test. Here are some examples:

-You might want to test whether the average age of clients engaging with the new process is the same as those engaging with the old process

-You might want to test if the average client tenure (how long they’ve been with Vanguard) of those engaging with the new process is the same as those engaging with the old process

-You might want to test if there are gender differences that affect engaging with the new or old process

Make sure to define the proper null and alternative hypothesis to test it. You are required to choose one or come up with another of your own to test.

In [198]:
test_group

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0          836976             6.0              73.0      60.5     U   
1          836976             6.0              73.0      60.5     U   
2          836976             6.0              73.0      60.5     U   
3          836976             6.0              73.0      60.5     U   
4          836976             6.0              73.0      60.5     U   
...           ...             ...               ...       ...   ...   
321190    7468138            18.0             222.0      61.0     F   
321191    7468138            18.0             222.0      61.0     F   
321192    7468138            18.0             222.0      61.0     F   
321193    7468138            18.0             222.0      61.0     F   
321194    7468138            18.0             222.0      61.0     F   

        num_accts        bal  calls_6_mnth  logons_6_mnth variation  \
0             2.0   45105.30           6.0            9.0      Test   
1             2.0   45105.30           6.0            9.0      Test   
2             2.0   45105.30           6.0            9.0      Test   
3             2.0   45105.30           6.0            9.0      Test   
4             2.0   45105.30           6.0            9.0      Test   
...           ...        ...           ...            ...       ...   
321190        3.0  209278.15           0.0            3.0      Test   
321191        3.0  209278.15           0.0            3.0      Test   
321192        3.0  209278.15           0.0            3.0      Test   
321193        3.0  209278.15           0.0            3.0      Test   
321194        3.0  209278.15           0.0            3.0      Test   

                   visitor_id                      visit_id process_step  \
0        427070339_1413275162   228976764_46825473280_96584      confirm   
1        427070339_1413275162   228976764_46825473280_96584      confirm   
2        427070339_1413275162   228976764_46825473280_96584      confirm   
3        427070339_1413275162   228976764_46825473280_96584       step_3   
4        427070339_1413275162   228976764_46825473280_96584       step_2   
...                       ...                           ...          ...   
321190  317123960_46820919455  769876461_30381166055_830233       step_2   
321191  317123960_46820919455  769876461_30381166055_830233       step_1   
321192  317123960_46820919455  769876461_30381166055_830233        start   
321193  317123960_46820919455  769876461_30381166055_830233        start   
321194  317123960_46820919455  769876461_30381166055_830233        start   

                 date_time  
0      2017-04-02 11:51:13  
1      2017-04-02 11:47:50  
2      2017-04-02 11:46:45  
3      2017-04-02 11:23:08  
4      2017-04-02 11:22:24  
...                    ...  
321190 2017-03-30 23:59:15  
321191 2017-03-30 23:58:51  
321192 2017-03-30 23:58:40  
321193 2017-03-30 23:55:11  
321194 2017-03-31 00:16:12  

[177787 rows x 14 columns]

In [200]:
# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'clnt_age': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'clnt_age': [...], 'process_step': [...]})

# Filter the data to get the client age for each group
test_age = test_group['clnt_age']
control_age = control_group['clnt_age']

# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_age, control_age, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in average age between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in average age between the Test and Control groups.")

Test Statistic: 7.928355666370457
P-value: 2.228041702345781e-15
Reject the null hypothesis: There is a significant difference in average age between the Test and Control groups.


In [199]:
# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'clnt_tenure_mnth': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'clnt_tenure_mnth': [...], 'process_step': [...]})

# Filter the data to get the client tenure for each group
test_tenure = test_group['clnt_tenure_mnth']
control_tenure = control_group['clnt_tenure_mnth']

# Perform the Two-sample t-test
stat, p_value = ttest_ind(test_tenure, control_tenure, equal_var=False)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Test Statistic: {stat}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in average client tenure between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in average client tenure between the Test and Control groups.")

Test Statistic: 0.505067363760497
P-value: 0.6135118864992013
Fail to reject the null hypothesis: There is no significant difference in average client tenure between the Test and Control groups.


In [202]:
from scipy.stats import chi2_contingency

# Assuming test_group and control_group DataFrames are already defined
# Sample data structure
# test_group = pd.DataFrame({'client_id': [...], 'gendr': [...], 'process_step': [...]})
# control_group = pd.DataFrame({'client_id': [...], 'gendr': [...], 'process_step': [...]})

# Create a contingency table for gender and group
contingency_table = pd.crosstab(tabla_analisis['variation'], tabla_analisis['gendr'])

# Perform the Chi-square test
chi2, p_value, dof, expected = chi2_contingency(contingency_table)

# Define the significance level
alpha = 0.05

# Print the results
print(f"Chi-square Statistic: {chi2}")
print(f"P-value: {p_value}")

if p_value < alpha:
    print("Reject the null hypothesis: There is a significant difference in gender proportions between the Test and Control groups.")
else:
    print("Fail to reject the null hypothesis: There is no significant difference in gender proportions between the Test and Control groups.")

Chi-square Statistic: 54.73241150360589
P-value: 7.830657522268704e-12
Reject the null hypothesis: There is a significant difference in gender proportions between the Test and Control groups.


### Conclusions:

Based on the results of the hypothesis tests, we can draw the following conclusions:

#### 1. Average Age of Clients:
- **Test Statistic**: 7.928355666370457
- **P-value**: 2.228041702345781e-15
- **Conclusion**: Reject the null hypothesis.

**Interpretation**:
- There is a statistically significant difference in the average age of clients between the Test and Control groups.
- This suggests that the new process (Test group) attracts clients of a different age profile compared to the old process (Control group).

#### 2. Average Client Tenure:
- **Test Statistic**: 0.505067363760497
- **P-value**: 0.6135118864992013
- **Conclusion**: Fail to reject the null hypothesis.

**Interpretation**:
- There is no statistically significant difference in the average client tenure between the Test and Control groups.
- This suggests that the new process (Test group) does not significantly differ from the old process (Control group) in terms of how long clients have been with Vanguard.

#### 3. Gender Differences:
- **Chi-square Statistic**: 54.73241150360589
- **P-value**: 7.830657522268704e-12
- **Conclusion**: Reject the null hypothesis.

**Interpretation**:
- There is a statistically significant difference in gender proportions between the Test and Control groups.
- This suggests that the new process (Test group) engages a different gender distribution compared to the old process (Control group).

### Practical Implications:
1. **Average Age**:
   - The significant difference in average age indicates that the new design may appeal more to a specific age group. This insight can be used to tailor marketing strategies and user experience improvements to better serve this demographic.

2. **Average Client Tenure**:
   - Since there is no significant difference in client tenure, the new design does not appear to affect how long clients have been with Vanguard. This suggests that the new design is equally engaging for both new and long-term clients.

3. **Gender Differences**:
   - The significant difference in gender proportions suggests that the new design may be more appealing to a particular gender. This insight can be used to further refine the design and marketing efforts to ensure inclusivity and appeal to a broader audience.

### Summary:
The hypothesis testing reveals significant differences in average age and gender proportions between the Test and Control groups, while there is no significant difference in average client tenure. These insights can guide further improvements and targeted strategies to enhance the effectiveness of the new design.

### Experiment Evaluation

#### Design Effectiveness

1. **Was the experiment well-structured?**
   - **Answer**: The experiment appears to be well-structured as it involves a clear comparison between the Test group (new design) and the Control group (old design). The use of statistical tests to compare completion rates, average age, client tenure, and gender proportions indicates a thorough approach to evaluating the effectiveness of the new design.

2. **Were clients randomly and equally divided between the old and new designs?**
   - **Answer**: To determine if clients were randomly and equally divided, we would need to verify the randomization process used in the experiment. If the assignment to Test and Control groups was done randomly and the sample sizes are large enough, we can assume that the division was fair. The provided data shows a substantial number of clients in both groups (177,787 in Test and 143,408 in Control), which suggests a reasonably balanced experiment.

3. **Were there any biases?**
   - **Answer**: Potential biases could include selection bias if the clients were not randomly assigned to the Test and Control groups. Additionally, if certain demographics (e.g., age, gender) were overrepresented in one group, it could introduce bias. The significant differences in average age and gender proportions between the groups suggest that there might have been some underlying biases in the assignment process. Ensuring proper randomization and stratification can help mitigate these biases.

#### Duration Assessment

1. **Was the timeframe of the experiment (from 3/15/2017 to 6/20/2017) adequate to gather meaningful data and insights?**
   - **Answer**: The experiment duration of approximately three months (from 3/15/2017 to 6/20/2017) seems adequate to gather meaningful data and insights. This timeframe allows for the collection of a substantial amount of data, as evidenced by the large sample sizes in both the Test and Control groups. It also provides enough time to observe user behavior and engagement with the new design.

#### Additional Data Needs

1. **What other data, if available, could enhance the analysis?**
   - **Answer**: Additional data that could enhance the analysis includes:
     - **Demographic Information**: More detailed demographic data (e.g., income level, education) could help identify specific segments of the population that respond differently to the new design.
     - **Behavioral Data**: Information on user interactions with the website (e.g., click-through rates, time spent on each page) could provide deeper insights into how the new design affects user behavior.
     - **Feedback and Satisfaction Scores**: Collecting user feedback and satisfaction scores could help understand the qualitative aspects of the user experience with the new design.
     - **Conversion Funnel Data**: Data on different stages of the conversion funnel (e.g., from initial visit to final conversion) could help identify where users drop off and how the new design impacts each stage.
     - **A/B Test Metadata**: Information on how the randomization was performed and any stratification criteria used could help assess the fairness and validity of the experiment.

### Summary

The experiment appears to be well-structured with a clear comparison between the Test and Control groups. The duration of the experiment seems adequate to gather meaningful data. However, potential biases in the assignment process should be addressed to ensure the validity of the results. Additional data on demographics, user behavior, feedback, and conversion funnel stages could further enhance the analysis and provide more comprehensive insights into the effectiveness of the new design.